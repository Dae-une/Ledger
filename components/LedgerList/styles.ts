import styled from '@emotion/styled';

export const Empty = styled.div`
  width: 100%;
  display: flex;
  margin-top: 5rem;
  justify-content: center;
  align-items: center;
`;

export const ListWrap = styled.div`
  padding-bottom: 6rem;
`;

export const ListStyle = styled.div`
  list-style: none;
  display: flex;
  justify-content: space-around;
  align-items: center;
  border-bottom: 1px solid rgb(228, 228, 228);
  height: 3rem;
  text-align: center;
  div {
    width: 100%;
    text-align: center;
  }
`;

export const Total = styled.div`
  width: 100%;
  display: flex;
  justify-content: space-evenly;
  height: 3rem;
  align-items: center;
`;
