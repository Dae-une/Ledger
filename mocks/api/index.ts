import { ScheduleType } from './../../types/types';
import axios, { AxiosRequestConfig } from 'axios';
import { LedgerType } from '../../types/types';

const axiosConfig: AxiosRequestConfig = {
  baseURL: 'http://localhost:3095',
};
const client = axios.create(axiosConfig);

export const getLedger = async (date: string) => {
  const { data } = await client.post<LedgerType[]>('/ledger', date);
  return data;
};
export const getSchedule = async (date: string) => {
  const { data } = await client.post<ScheduleType[]>('/schedule', date);
  return data;
};
