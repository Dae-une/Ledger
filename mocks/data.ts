export const ledgerData = [
  {
    type: '식사',
    desc: 'OO과 식사',
    price: 14000,
    date: '2023-01-04',
  },
  {
    type: '구매',
    desc: '코딩 책 구매',
    price: 14000,
    date: '2023-01-05',
  },
  {
    type: '저축',
    desc: '적금 ',
    price: 50000,
    date: '2023-01-04',
  },
  {
    type: '식사',
    desc: 'OO과 식사',
    price: 14000,
    date: '2023-01-05',
  },
  {
    type: '식사',
    desc: 'OO과 식사',
    price: 14000,
    date: '2023-01-04',
  },
  {
    type: '식사',
    desc: 'OO과 식사',
    price: 14000,
    date: '2023-01-01',
  },
  {
    type: '식사',
    desc: 'OO과 식사',
    price: 14000,
    date: '2023-01-18',
  },
];

export const scheduleData = [
  {
    id: 1,
    title: '제목',
    desc: '무슨일인지 쥬르르르르르르르르륵',
    startDate: '2023-01-14',
    endDate: '2023-01-14',
  },
  {
    id: 1,
    title: '제목2',
    desc: '무슨일인지 쥬르르르르르르르르륵2',
    startDate: '2023-01-14',
    endDate: '2023-01-14',
  },
  {
    id: 1,
    title: '제목3',
    desc: '무슨일인지 쥬르르르르르르르르륵3',
    startDate: '2023-01-14',
    endDate: '2023-01-14',
  },
  {
    id: 1,
    title: '제목4',
    desc: '무슨일인지 쥬르르르르르르르르륵2',
    startDate: '2023-01-14',
    endDate: '2023-01-14',
  },
  {
    title: '제목5',
    desc: '무슨일인지 쥬르르르르르르르르륵1',
    startDate: '2023-01-14',
    endDate: '2023-01-14',
  },
  {
    title: '제목',
    desc: '무슨일인지 쥬르르르르르르르르륵',
    startDate: '2023-01-14',
    endDate: '2023-01-14',
  },
  {
    title: '제목5',
    desc: '무슨일인지 쥬르르르르르르르르륵1',
    startDate: '2023-01-14',
    endDate: '2023-01-14',
  },
  {
    title: '제목',
    desc: '무슨일인지 쥬르르르르르르르르륵',
    startDate: '2023-01-14',
    endDate: '2023-01-14',
  },
];
