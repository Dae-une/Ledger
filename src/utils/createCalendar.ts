import { Dayjs } from 'dayjs';

const createCalendar = (baseDate: Dayjs) => {
  const monthList = [];
  const WEEK = 7;
  const monthStart = baseDate.startOf('month');
  const monthEnd = baseDate.endOf('month');
  const startDay = monthStart.day();
  const endDate = monthEnd.date();
  const totalWeek = Math.ceil((startDay + endDate) / 7);

  for (let i = 1; i <= totalWeek; i++) {
    const weekList = [];
    if (i === 1) {
      const prevMonth = baseDate.subtract(1, 'month');
      for (let j = 1; j <= WEEK; j++) {
        let date;
        if (j <= startDay) {
          date = prevMonth.endOf('month').subtract(startDay - j, 'day');
        } else {
          date = monthStart.add(j - 1 - startDay, 'day');
        }
        weekList.push(date);
      }
    } else {
      const startDate = (i - 1) * WEEK;
      for (let j = startDate; j <= i * WEEK - 1; j++) {
        let date;
        if (j - startDay < endDate) {
          date = monthStart.add(j - startDay, 'day');
        } else {
          date = monthEnd.add(j - endDate - startDay + 1, 'day');
        }
        weekList.push(date);
      }
    }
    monthList.push(weekList);
  }

  return monthList;
};

export default createCalendar;
