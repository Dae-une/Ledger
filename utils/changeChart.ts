import { LedgerType } from '../types/types';

interface ChartData {
  id: string;
  value: number;
}

export const chageToChart = (data: LedgerType[]) => {
  const chartData: ChartData[] = [];
  data.map((list) => {
    const index = chartData.findIndex((data) => data.id === list.type);
    if (index < 0) {
      chartData.push({ id: list.type, value: list.price });
    } else {
      chartData[index].value += list.price;
    }
  });

  return chartData;
};
