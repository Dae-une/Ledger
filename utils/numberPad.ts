export const numberPad = (n: number, width: number) => {
  const v = n + '';
  return v.length >= width ? n : new Array(width - v.length + 1).join('0') + v;
};
