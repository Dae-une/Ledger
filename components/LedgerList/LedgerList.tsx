import React from 'react';
import { LedgerType } from '../../types/types';
import changeInPrice from '../../utils/changeInPrice';
import { Empty, ListStyle, ListWrap } from './styles';

interface Props {
  list: LedgerType[] | undefined;
}

const LedgerList = ({ list }: Props) => {
  if (!list?.length) {
    return <Empty>내역이 없어요.</Empty>;
  }
  return (
    <ListWrap>
      <ListStyle>
        <div>타입</div>
        <div>내역</div>
        <div>금액</div>
      </ListStyle>
      {list.map((data) => (
        <ListStyle key={data.desc}>
          <div>{data.type}</div>
          <div>{data.desc}</div>
          <div>{changeInPrice(data.price)}원</div>
        </ListStyle>
      ))}
    </ListWrap>
  );
};

export default LedgerList;
