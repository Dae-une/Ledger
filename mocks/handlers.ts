import { rest } from 'msw';
import { ledgerData, scheduleData } from './data';

export const handlers = [
  rest.post('http://localhost:3095/ledger', async (req, res, ctx) => {
    const data = await req.text();
    const [year, month, date] = data.split('-');
    const returnData = ledgerData.filter((ledger) => {
      const [dYear, dMonth, dDate] = ledger.date.split('-');
      return dYear === year && month === dMonth;
    });

    if (!returnData.length) {
      return res(ctx.status(401));
    }
    return res(ctx.status(200), ctx.json(returnData));
  }),

  rest.post('http://localhost:3095/schedule', async (req, res, ctx) => {
    const data = await req.text();
    const [year, month, date] = data.split('-');

    const returnData = scheduleData.filter((schedule) => {
      const [dYear, dMonth, dDate] = schedule.date.split('-');
      return dYear === year && month === dMonth;
    });

    if (!returnData.length) {
      return res(ctx.status(401));
    }
    return res(ctx.status(200), ctx.json(returnData));
  }),
];
