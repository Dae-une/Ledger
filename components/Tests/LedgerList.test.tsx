import React from 'react';
import { screen, render, waitForElementToBeRemoved } from '@testing-library/react';
import { Route } from 'react-router-dom';
import { ledgerData } from '../../mocks/data';
import { withRouter } from '../../tests/utils';
import LedgerList from '../LedgerList/LedgerList';

describe('LedgerList', () => {
  it('render correctly', async () => {
    const { asFragment } = render(withRouter(<Route path="/" element={<LedgerList list={ledgerData} />} />));

    expect(asFragment()).toMatchSnapshot();
  });

  it('리스트가 없으면 Empty화면을 보여준다.', () => {
    render(withRouter(<Route path="/" element={<LedgerList list={[]} />} />));

    expect(screen.getByText('내역이 없어요.'));
  });
});
