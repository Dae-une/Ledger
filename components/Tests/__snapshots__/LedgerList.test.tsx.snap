// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LedgerList render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-rjennl"
  >
    <div
      class="css-sofpfz"
    >
      <div>
        타입
      </div>
      <div>
        내역
      </div>
      <div>
        금액
      </div>
    </div>
    <div
      class="css-sofpfz"
    >
      <div>
        식사
      </div>
      <div>
        OO과 식사
      </div>
      <div>
        14,000원
      </div>
    </div>
    <div
      class="css-sofpfz"
    >
      <div>
        구매
      </div>
      <div>
        코딩 책 구매
      </div>
      <div>
        14,000원
      </div>
    </div>
    <div
      class="css-sofpfz"
    >
      <div>
        저축
      </div>
      <div>
        적금 
      </div>
      <div>
        50,000원
      </div>
    </div>
    <div
      class="css-sofpfz"
    >
      <div>
        식사
      </div>
      <div>
        OO과 식사
      </div>
      <div>
        14,000원
      </div>
    </div>
    <div
      class="css-sofpfz"
    >
      <div>
        식사
      </div>
      <div>
        OO과 식사
      </div>
      <div>
        14,000원
      </div>
    </div>
    <div
      class="css-sofpfz"
    >
      <div>
        식사
      </div>
      <div>
        OO과 식사
      </div>
      <div>
        14,000원
      </div>
    </div>
    <div
      class="css-sofpfz"
    >
      <div>
        식사
      </div>
      <div>
        OO과 식사
      </div>
      <div>
        14,000원
      </div>
    </div>
    <div
      class="css-vuuins"
    >
      <span>
        총 금액 
      </span>
      <span>
        134,000원
      </span>
    </div>
  </div>
</DocumentFragment>
`;
